db/memorystore.go

```diff
@@ -170,6 +172,13 @@ func (m *MemoryStore) AddOrder(order *core.Order) (int, error) {
 		return 0, fmt.Errorf("order %q already exists", orderID)
 	}
 
+	var ordersByAccountID []*core.Order
+	var present bool
+	if ordersByAccountID, present = m.ordersByAccountID[order.AccountID]; !present {
```

Comment:
I think you need to capture this `order.AccountID` at the same place the `orderID` is assigned in order to do it while the `order` is `RLock()`'d

Reply:
Good point!

Reply:
I've also moved the lock release before the `if`, so that the lock is released in any case.